### ISSUE TITLE
[QST] How to link the cudf lib in a C++ project?

### ISSUE THREAD

**Issue by ChenZhongPu** [rapidsai/cudf#7566]
I would like to run the example C++ code shown in [libcudf api](https://docs.rapids.ai/api/libcudf/nightly/group__io__readers.html#ga9d0094c2a97b3568e80209588644b38c).

```c++
std::string filepath = "tips.csv";
auto source_info = cudf::io::source_info(filepath); // ok
auto builder = cudf::io::csv_reader_options::builder(source_info); // error
```
The second line is ok, but the third line would cause the error by `nvcc  -I/home/zhongpu/miniconda3/envs/cudf_dev/include -L/home/zhongpu/miniconda3/envs/cudf_dev/lib test.cu`

> /usr/bin/ld: /tmp/tmpxft_0000b602_00000000-11_test.o: in function `main':
> tmpxft_0000b602_00000000-6_test.cudafe1.cpp:(.text+0xf9): undefined reference to `cudf::io::csv_reader_options::builder(cudf::io::source_info const&)'
collect2: error: ld returned 1 exit status

I think this is caused by the link issue, because `nvcc  -I/home/zhongpu/miniconda3/envs/cudf_dev/include -c test.cu` works.

The complete `test.cu`:

```c++
#include <iostream>
#include <string>
#include <cudf/column/column.hpp>
#include <cudf/io/types.hpp>
#include <cudf/io/csv.hpp>

int main() {
    std::string filepath = "/home/zhongpu/Download/tips.csv";
    auto source_info = cudf::io::source_info(filepath);
    auto builder = cudf::io::csv_reader_options::builder(source_info);   
}
```

**Comment by davidwendt** [comment]
Which version of libcudf? 
You probably need to link to `libcudf.so` Can you try adding '-lcudf' ?
`nvcc -I/home/zhongpu/miniconda3/envs/cudf_dev/include -L/home/zhongpu/miniconda3/envs/cudf_dev/lib -lcudf test.cu`

**Comment by ChenZhongPu** [comment]
Yes, it is necessary to add `-lcudf`.

But another issue found in this small code snippet. 

First of all, there is a mistake in the sample code `auto result = cudf::read_csv(options)`, right? Because `read_csv` is in the namespace of `cudf::io`.

Second, it would throw an error:

>/usr/include/c++/9/bits/unique_ptr.h(79): error: incomplete type is not allowed
          detected during:
            instantiation of "void std::default_delete<_Tp>::operator()(_Tp *) const [with _Tp=cudf::table]" 
(292): here
            instantiation of "std::unique_ptr<_Tp, _Dp>::~unique_ptr() [with _Tp=cudf::table, _Dp=std::default_delete<cudf::table>]" 
/home/zhongpu/miniconda3/envs/cudf_dev/include/cudf/io/types.hpp(159): here

1 error detected in the compilation of "test.cu".


My build is based on the `commit 850548d` of branch 0.19.

```c++
#include <iostream>
#include <string>
#include <cudf/column/column.hpp>
#include <cudf/io/types.hpp>
#include <cudf/io/csv.hpp>

int main() {
    std::string filepath = "/home/zhongpu/Download/tips.csv";
    auto source_info = cudf::io::source_info(filepath);
    cudf::io::csv_reader_options options = cudf::io::csv_reader_options::builder(source_info);   
    auto result = cudf::io::read_csv(options);
}
```

**Comment by jrhemstad** [comment]
It is strongly recommended that you link against the CMake target that is exposed in the libcudf cmake file. We are currently working on an example repo to show how to do this. You can see the in progress repo here for inspiration: https://github.com/isVoid/linkto_libcudf

**Comment by ChenZhongPu** [comment]
Yes, it is better to link against CMake target, but I still found something that it is hard for me to understand. So, I hope anyone can help me figure this out.

1. `nvcc -I/home/zhongpu/miniconda3/envs/cudf_dev/include -L/home/zhongpu/miniconda3/envs/cudf_dev/lib -lcudf test.cu` would still throw error, and I also have to link `-lboost_filesystem -larrow -lnvrtc` to generate a executable file.

To check the its dependent shared libs, `ldd a.out`: (Note: **libcudf.so => not found**)

> 	linux-vdso.so.1 (0x00007ffd28c7f000)
	libcudf.so => not found
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f648a405000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f648a3e2000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f648a3dc000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f648a1fb000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f648a1e0000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f6489fec000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f648a4ce000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f6489e9d000)

So, I have to set `LD_LIBRARY_PATH` to `/home/zhongpu/miniconda3/envs/cudf_dev/lib`.  

2. As long as I set `LD_LIBRARY_PATH`, I found that I am able to `nvcc -I/home/zhongpu/miniconda3/envs/cudf_dev/include -L/home/zhongpu/miniconda3/envs/cudf_dev/lib -lcudf test.cu` directly (without additional `boost_filesystem`, `arrow` and `nvrtc`). From my perspective, `LD_LIBRARY_PATH` is **only** used for program executing, so why is it affecting the compiling and linking stages? 

3. I try to add `-Xlinker -rpath` as `nvcc` options (without setting `LD_LIBRARY_PATH`): `nvcc -I/home/zhongpu/miniconda3/envs/cudf_dev/include -L/home/zhongpu/miniconda3/envs/cudf_dev/lib -lcudf -Xlinker "-rpath /home/zhongpu/miniconda3/envs/cudf_dev/lib" test.cu`. It works, but `ldd a.out` shows: (note: **`boost_filesystem`, `arrow` and `nvrtc` are not found**)

> 	linux-vdso.so.1 (0x00007ffe7eb59000)
	libcudf.so => /home/zhongpu/miniconda3/envs/cudf_dev/lib/libcudf.so (0x00007f9b3d2b9000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f9b3d292000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f9b3d26f000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f9b3d269000)
	libstdc++.so.6 => /home/zhongpu/miniconda3/envs/cudf_dev/lib/libstdc++.so.6 (0x00007f9b3d0f4000)
	libgcc_s.so.1 => /home/zhongpu/miniconda3/envs/cudf_dev/lib/libgcc_s.so.1 (0x00007f9b3d0e0000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f9b3ceec000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9b524b9000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f9b3ced0000)
	libboost_filesystem.so.1.75.0 => not found
	libarrow.so.100 => not found
	libcudart.so.11.0 => /usr/local/cuda-11.2/lib64/libcudart.so.11.0 (0x00007f9b3cc41000)
	libcuda.so.1 => /usr/lib/x86_64-linux-gnu/libcuda.so.1 (0x00007f9b3b55b000)
	libnvrtc.so.11.0 => not found
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f9b3b40a000)

It seems that I still have to add `boost_filesystem`, `arrow` and `nvrtc`, but even if I use `nvcc -I/home/zhongpu/miniconda3/envs/cudf_dev/include -L/home/zhongpu/miniconda3/envs/cudf_dev/lib -lcudf -lboost_filesystem -larrow -lnvrtc -Xlinker "-rpath /home/zhongpu/miniconda3/envs/cudf_dev/lib" test.cu`, the output of `ldd a.out` remains unchanged. 

In other words, I still have to set `LD_LIBRARY_PATH` manually. How to explain this? And is it possible to run this executable file without setting  `LD_LIBRARY_PATH`?
